scipy/optimize/_minpack_py.py

```diff
@@ -889,7 +923,7 @@ def curve_fit(f, xdata, ydata, p0=None, sigma=None, absolute_sigma=False,
         pcov = np.dot(VT.T / s**2, VT)
 
     warn_cov = False
-    if pcov is None:
+    if pcov is None or np.isnan(pcov).any():
```

Comment:
Not sure, but maybe in PR #17247, the `leastsq` covariance calculation was changed. It started to return a covariance matrix including nans, instead of returning None. So, I added this.

Reply:
This is a bug fix, so I think we can backport it.

Reply:
@AtsushiSakai I'm not able to reproduce what you're describing. For instance:

```python3
from scipy import optimize
def f(x, a, b):
    return a + b

x = [1, 2, 3, 4]
y = [1, 1, 1, 1]
res = optimize.curve_fit(f, x, y)
res
```
gives
```
(array([-2.18114415e-12,  1.00000000e+00]),
 array([[inf, inf],
        [inf, inf]]))
```
The relevant code is:
https://github.com/mdhaber/scipy/blob/e099c4d2939d1f1c7201d542096eb2cbf6c7da92/scipy/optimize/_minpack_py.py#L467-L488

I don't know how to get NaNs; the `LinAlgError` seems to get raised when there are problems, and causes `cov_x` returned by `leastsq` to stay `None` and `curve_fit` to return an array of `inf`s.

Do you have an example?

Reply:
@mdhaber This is an example:

Main branch:
```
In [10]: import numpy as np
In [11]: from scipy import optimize
In [12]: xdata_with_nan = np.array([2, 3, np.nan, 4, 4, np.nan])
In [13]: ydata_with_nan = np.array([1, 2, 5, 3, np.nan, 7])
In [14]: optimize.curve_fit(f, xdata_with_nan, ydata_with_nan, check_finite=False)
Out[14]: 
(array([1., 1.]),
 array([[nan, nan],
        [nan, nan]]))
```

Scipy 1.9.3
```
In [10]: import numpy as np
In [11]: from scipy import optimize
In [12]: xdata_with_nan = np.array([2, 3, np.nan, 4, 4, np.nan])
In [13]: ydata_with_nan = np.array([1, 2, 5, 3, np.nan, 7])
In [14]: optimize.curve_fit(f, xdata_with_nan, ydata_with_nan, check_finite=False)
optimize.curve_fit(f, xdata_with_nan, ydata_with_nan, check_finite=False
   ...: )
/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages/scipy/optimize/_minpack_py.py:881: OptimizeWarning: Covariance of the parameters could not be estimated
  warnings.warn('Covariance of the parameters could not be estimated',
Out[8]:
(array([1., 1.]),
 array([[inf, inf],
        [inf, inf]]))
```

Reply:
If this can only reproduced by including NaNs in the data, this does not need to be backported because NaNs are currently not supported input. 
It does not even need to be changed here. Here are the three possibilities:

- With `nan_policy='raise'`, the function will error out before getting this far.
- With `nan_policy='omit'`, the NaNs will be omitted and this won't happen.
- With `nan_policy=None`, the behavior is said to be implementation-dependent and not guaranteed. 

Personally, I prefer the original behavior for `nan_policy=None`.
```suggestion
    if pcov is None:
```
Although the behavior doesn't really matter, I think an array of NaNs will usually be less incorrect than an array of `inf`s in case of NaN input.